File: python/src/causentra/adapter.py
"""Safe bridge for framework lifecycle callbacks."""

from __future__ import annotations

import hashlib
import secrets
import threading
from collections.abc import Callable
from contextlib import suppress
from datetime import datetime, timezone

from .redaction import default_redactor
from .types import (
    SCHEMA_VERSION,
    EventAttributes,
    EventExporter,
    EventStatus,
    Redactor,
    RuntimeErrorContext,
    RuntimeEvent,
    TraceContext,
)
from .validation import validate_attributes, validate_event


class AdapterEventBridge:
    """Normalize unstable external IDs into the Causentra v1 contract."""

    def __init__(
        self,
        service_name: str,
        exporter: EventExporter,
        *,
        redactor: Redactor = default_redactor,
        on_error: Callable[[RuntimeErrorContext], None] | None = None,
        parent_context: TraceContext | None = None,
    ) -> None:
        if not service_name.strip():
            raise ValueError("service_name must not be empty")
        self._service_name = service_name
        self._exporter = exporter
        self._redactor = redactor
        self._on_error = on_error or (lambda _context: None)
        self._parent_context = parent_context
        self._sequences: dict[str, int] = {}
        self._lock = threading.Lock()

    def emit(
        self,
        *,
        external_trace_id: str,
        external_span_id: str,
        event_type: str,
        name: str,
        attributes: EventAttributes | None = None,
        external_parent_span_id: str | None = None,
        external_session_id: str | None = None,
        status: EventStatus = "unset",
        timestamp: datetime | None = None,
        duration_ms: float | None = None,
    ) -> bool:
        try:
            trace_id = (
                self._parent_context.trace_id
                if self._parent_context is not None
                else _canonical(external_trace_id, 32)
            )
            span_id = _canonical(external_span_id, 16)
            if external_parent_span_id is not None:
                parent_id = _canonical(external_parent_span_id, 16)
            elif self._parent_context is not None:
                parent_id = self._parent_context.span_id[-16:]
            else:
                parent_id = None
            raw = attributes or {}
            validate_attributes(raw)
            safe = self._redactor(raw)
            validate_attributes(safe)
            when = timestamp or datetime.now(timezone.utc)
            event = RuntimeEvent(
                schema_version=SCHEMA_VERSION,
                event_id=secrets.token_hex(16),
                trace_id=trace_id,
                span_id=span_id,
                parent_span_id=parent_id,
                session_id=(
                    _canonical(external_session_id, 32)
                    if external_session_id
                    else self._parent_context.session_id
                    if self._parent_context
                    else None
                ),
                sequence=self._next(trace_id),
                timestamp=when.astimezone(timezone.utc)
                .isoformat(timespec="milliseconds")
                .replace("+00:00", "Z"),
                type=event_type,
                name=name[:256],
                status=status,
                duration_ms=duration_ms,
                service_name=self._service_name,
                attributes=safe,
            )
            validate_event(event)
            self._exporter.emit(event)
            return True
        except BaseException as error:
            with suppress(BaseException):
                self._on_error(RuntimeErrorContext("adapter", error, 1))
            return False

    def flush(self, timeout: float | None = None) -> bool:
        return self._exporter.flush(timeout)

    def shutdown(self, timeout: float | None = None) -> None:
        self._exporter.shutdown(timeout)

    def _next(self, trace_id: str) -> int:
        with self._lock:
            value = self._sequences.get(trace_id, 0)
            self._sequences[trace_id] = value + 1
            return value


def _canonical(value: str, length: int) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()[:length]

File: packages/sdk/src/adapter.ts
import { createHash, randomBytes } from "node:crypto";
import { defaultRedactor } from "./redaction.js";
import {
  SCHEMA_VERSION,
  type EventAttributes,
  type EventExporter,
  type EventStatus,
  type EventType,
  type Redactor,
  type RuntimeErrorContext,
  type RuntimeEvent,
  type TraceContext,
} from "./types.js";
import { validateAttributes, validateEvent } from "./validation.js";

/** Configuration shared by framework adapters. */
export interface AdapterEventBridgeOptions {
  readonly serviceName: string;
  readonly exporter: EventExporter;
  readonly redactor?: Redactor;
  readonly onError?: (context: RuntimeErrorContext) => void;
  readonly now?: () => Date;
  /** Existing canonical parent used to join this adapter to a larger trace. */
  readonly parentContext?: TraceContext;
}

/** Framework lifecycle input converted to the stable runtime envelope. */
export interface AdapterEventInput {
  readonly externalTraceId: string;
  readonly externalSpanId: string;
  readonly externalParentSpanId?: string;
  readonly externalSessionId?: string;
  readonly timestamp?: string;
  readonly type: EventType;
  readonly name: string;
  readonly status?: EventStatus;
  readonly durationMs?: number;
  readonly attributes?: EventAttributes;
}

/**
 * Safe adapter-authoring surface. It converts framework identifiers into
 * deterministic Causentra IDs, applies redaction, assigns trace-local
 * sequence numbers, validates the result, and emits through a public exporter.
 * Framework packages never need access to SDK internals.
 */
export class AdapterEventBridge {
  readonly #serviceName: string;
  readonly #exporter: EventExporter;
  readonly #redactor: Redactor;
  readonly #onError: (context: RuntimeErrorContext) => void;
  readonly #now: () => Date;
  readonly #parentContext: TraceContext | undefined;
  readonly #sequences = new Map<string, number>();

  public constructor(options: AdapterEventBridgeOptions) {
    if (options.serviceName.trim().length === 0) {
      throw new TypeError("serviceName must not be empty");
    }
    this.#serviceName = options.serviceName;
    this.#exporter = options.exporter;
    this.#redactor = options.redactor ?? defaultRedactor;
    this.#onError = options.onError ?? (() => undefined);
    this.#now = options.now ?? (() => new Date());
    this.#parentContext = options.parentContext;
    if (this.#parentContext !== undefined) validateParentContext(this.#parentContext);
  }

  /** Emits one framework lifecycle event; returns undefined when safely dropped. */
  public emit(input: AdapterEventInput): RuntimeEvent | undefined {
    try {
      requireExternalId(input.externalTraceId, "externalTraceId");
      requireExternalId(input.externalSpanId, "externalSpanId");
      const traceId = this.#parentContext?.traceId
        ?? normalizeExternalId("trace", input.externalTraceId);
      const attributes = input.attributes ?? {};
      validateAttributes(attributes);
      const redacted = this.#redactor(attributes);
      validateAttributes(redacted);
      const event: RuntimeEvent = {
        schemaVersion: SCHEMA_VERSION,
        eventId: randomBytes(16).toString("hex"),
        traceId,
        spanId: normalizeExternalId(
          "span",
          `${input.externalTraceId}\0${input.externalSpanId}`,
        ),
        ...parentSpan(input, this.#parentContext),
        ...session(input, this.#parentContext),
        sequence: this.#nextSequence(traceId),
        timestamp: input.timestamp ?? this.#now().toISOString(),
        type: input.type,
        name: input.name,
        status: input.status ?? "unset",
        ...(input.durationMs === undefined ? {} : { durationMs: input.durationMs }),
        serviceName: this.#serviceName,
        attributes: redacted,
      };
      validateEvent(event);
      this.#exporter.emit(event);
      if (input.type === "trace.end") this.#sequences.delete(traceId);
      return event;
    } catch (error) {
      this.#report(error);
      return undefined;
    }
  }

  public flush(): Promise<boolean> {
    return this.#exporter.flush();
  }

  public async shutdown(): Promise<void> {
    if (this.#exporter.shutdown !== undefined) await this.#exporter.shutdown();
    else await this.#exporter.flush();
  }

  #nextSequence(traceId: string): number {
    const sequence = this.#sequences.get(traceId) ?? 0;
    this.#sequences.set(traceId, sequence + 1);
    return sequence;
  }

  #report(error: unknown): void {
    try {
      this.#onError({ operation: "adapter", error, droppedEvents: 1 });
    } catch {
      // Adapter diagnostics are application callbacks and cannot be trusted.
    }
  }
}

function parentSpan(
  input: AdapterEventInput,
  parent: TraceContext | undefined,
): { readonly parentSpanId?: string } {
  if (input.externalParentSpanId !== undefined) {
    return {
      parentSpanId: normalizeExternalId(
        "span",
        `${input.externalTraceId}\0${input.externalParentSpanId}`,
      ),
    };
  }
  return parent === undefined ? {} : { parentSpanId: parent.spanId };
}

function session(
  input: AdapterEventInput,
  parent: TraceContext | undefined,
): { readonly sessionId?: string } {
  if (input.externalSessionId !== undefined) {
    return { sessionId: normalizeExternalId("session", input.externalSessionId) };
  }
  return parent?.sessionId === undefined ? {} : { sessionId: parent.sessionId };
}

function validateParentContext(context: TraceContext): void {
  if (!/^[0-9a-f]{32}$/u.test(context.traceId)) {
    throw new TypeError("parentContext.traceId must be a 32-character lowercase hex ID");
  }
  if (!/^(?:[0-9a-f]{16}|[0-9a-f]{32})$/u.test(context.spanId)) {
    throw new TypeError("parentContext.spanId must be a 16- or 32-character lowercase hex ID");
  }
  if (context.sessionId !== undefined && !/^[0-9a-f]{32}$/u.test(context.sessionId)) {
    throw new TypeError("parentContext.sessionId must be a 32-character lowercase hex ID");
  }
}

/** Deterministically maps a framework identifier into a lowercase 128-bit ID. */
export function normalizeExternalId(namespace: string, value: string): string {
  return createHash("sha256")
    .update(namespace)
    .update("\0")
    .update(value)
    .digest("hex")
    .slice(0, namespace === "span" ? 16 : 32);
}

function requireExternalId(value: string, field: string): void {
  if (typeof value !== "string" || value.length === 0 || value.length > 1_024) {
    throw new TypeError(`${field} must be a non-empty string up to 1024 characters`);
  }
}
